Issue with the user creation with group and role assignments

I am creating a user that gets assigned to a group and the role gets assigned to the group. It works great when I run it once. However if I run it second time, it gives me the following error

Error: [ERROR] Error Updating Admin Roles On User: The API returned an error: Not found: Resource not found: gramupp46qf7QZyxU0h7 (RoleAssignment)

 180: resource okta_user admin_user1{

I have something like this:

resource okta_group super_admin_group {
  name        = "Super Admin"
  description = "Super Admin Group"
}

resource okta_group_roles  super_admin_group_role {
  group_id    = "${okta_group.super_admin_group.id}"
  admin_roles = ["SUPER_ADMIN"]
}


resource okta_user admin_user1{
  first_name = "Joe"
  last_name  = "Smith"
  login      = "joe.smith@company.com"
  email      = "joe.smith@company.com"
  group_memberships = ["${okta_group.super_admin_group.id}"]
  depends_on = [okta_group.super_admin_group, okta_group_roles.super_admin_group_role]
}

--- a/examples/okta_group_roles/all_roles.tf
+++ b/examples/okta_group_roles/all_roles.tf
@@ -17,3 +17,11 @@ resource okta_group_roles test {
     "HELP_DESK_ADMIN",
   ]
 }
+
+resource okta_user test {
+  first_name        = "TestAcc"
+  last_name         = "Smith"
+  login             = "test-acc-replace_with_uuid@testing.com"
+  email             = "test-acc-replace_with_uuid@testing.com"
+  group_memberships = ["${okta_group.test.id}"]
+}

--- a/examples/okta_group_roles/basic.tf
+++ b/examples/okta_group_roles/basic.tf
@@ -7,3 +7,11 @@ resource okta_group_roles test {
   group_id    = "${okta_group.test.id}"
   admin_roles = ["SUPER_ADMIN"]
 }
+
+resource okta_user test {
+  first_name        = "TestAcc"
+  last_name         = "Smith"
+  login             = "test-acc-replace_with_uuid@testing.com"
+  email             = "test-acc-replace_with_uuid@testing.com"
+  group_memberships = ["${okta_group.test.id}"]
+}

--- a/examples/okta_user_base_schema/basic.tf
+++ b/examples/okta_user_base_schema/basic.tf
@@ -1,7 +1,7 @@
 resource "okta_user_base_schema" "firstName" {
-  index = "firstName"
-  title = "First name"
-  type  = "string"
-
+  index       = "firstName"
+  master      = "PROFILE_MASTER"
   permissions = "READ_ONLY"
+  title       = "First name"
+  type        = "string"
 }

--- a/examples/okta_user_base_schema/updated.tf
+++ b/examples/okta_user_base_schema/updated.tf
@@ -1,7 +1,7 @@
 resource "okta_user_base_schema" "firstName" {
-  index = "firstName"
-  title = "First name"
-  type  = "string"
-
+  index       = "firstName"
+  title       = "First name"
+  type        = "string"
+  master      = "PROFILE_MASTER"
   permissions = "READ_WRITE"
 }

--- a/go.mod
+++ b/go.mod
@@ -24,4 +24,4 @@ require (
 	golang.org/x/sys v0.0.0-20190528012530-adf421d2caf4 // indirect
 )
 
-replace github.com/okta/okta-sdk-golang => github.com/articulate/okta-sdk-golang v1.0.5
+replace github.com/okta/okta-sdk-golang => github.com/articulate/okta-sdk-golang v0.0.0-20190810203837-596e830a6cb1

--- a/okta/app.go
+++ b/okta/app.go
@@ -332,7 +332,7 @@ func handleAppUsers(id string, d *schema.ResourceData, client *okta.Client) []fu
 			if !contains(userIDList, user.Id) {
 				userID := user.Id
 				asyncActionList = append(asyncActionList, func() error {
-					return suppressErrorOn404(client.Application.DeleteApplicationUser(id, userID))
+					return suppressErrorOn404(client.Application.DeleteApplicationUser(id, userID, nil))
 				})
 			}
 		}

--- a/okta/config.go
+++ b/okta/config.go
@@ -1,6 +1,7 @@
 package okta
 
 import (
+	"context"
 	"fmt"
 	"net/http"
 	"time"
@@ -9,7 +10,6 @@ import (
 	"github.com/hashicorp/go-cleanhttp"
 	"github.com/hashicorp/terraform/helper/logging"
 	"github.com/okta/okta-sdk-golang/okta"
-	"github.com/okta/okta-sdk-golang/okta/cache"
 )
 
 func (adt *AddHeaderTransport) RoundTrip(req *http.Request) (*http.Response, error) {
@@ -56,20 +56,25 @@ func (c *Config) loadAndValidate() error {
 
 	orgUrl := fmt.Sprintf("https://%v.%v", c.orgName, c.domain)
 
-	config := okta.NewConfig().
-		WithOrgUrl(orgUrl).
-		WithToken(c.apiToken).
-		WithCache(false).
-		WithBackoff(c.backoff).
-		WithMinWait(time.Duration(c.minWait) * time.Second).
-		WithMaxWait(time.Duration(c.maxWait) * time.Second).
-		WithRetries(int32(c.retryCount))
-	client := okta.NewClient(config, httpClient, cache.NewNoOpCache())
+	client, err := okta.NewClient(
+		context.Background(),
+		okta.WithOrgUrl(orgUrl),
+		okta.WithToken(c.apiToken),
+		okta.WithCache(false),
+		okta.WithBackoff(c.backoff),
+		okta.WithMinWait(time.Duration(c.minWait)*time.Second),
+		okta.WithMaxWait(time.Duration(c.maxWait)*time.Second),
+		okta.WithRetries(int32(c.retryCount)),
+		okta.WithHttpClient(*httpClient),
+	)
+	if err != nil {
+		return err
+	}
 	c.supplementClient = &ApiSupplement{
 		baseURL:         fmt.Sprintf("https://%s.%s", c.orgName, c.domain),
 		client:          httpClient,
 		token:           c.apiToken,
-		requestExecutor: okta.NewRequestExecutor(httpClient, cache.NewNoOpCache(), config),
+		requestExecutor: client.GetRequestExecutor(),
 	}
 
 	// add the Okta SDK client object to Config

--- a/okta/provider_sweeper_test.go
+++ b/okta/provider_sweeper_test.go
@@ -1,6 +1,7 @@
 package okta
 
 import (
+	"context"
 	"fmt"
 	"strconv"
 	"testing"
@@ -85,8 +86,16 @@ func sharedClient(region string) (*articulateOkta.Client, *okta.Client, *ApiSupp
 
 	orgURL := fmt.Sprintf("https://%v.%v", c.orgName, c.domain)
 
-	config := okta.NewConfig().WithOrgUrl(orgURL).WithToken(c.apiToken).WithBackoff(true).WithRetries(20)
-	client := okta.NewClient(config, nil, nil)
+	client, err := okta.NewClient(
+		context.Background(),
+		okta.WithOrgUrl(orgURL),
+		okta.WithToken(c.apiToken),
+		okta.WithBackoff(true),
+		okta.WithRetries(20),
+	)
+	if err != nil {
+		return articulateClient, client, nil, err
+	}
 	api := &ApiSupplement{requestExecutor: client.GetRequestExecutor()}
 
 	return articulateClient, client, api, nil

--- a/okta/resource_user.go
+++ b/okta/resource_user.go
@@ -398,7 +398,7 @@ func resourceUserUpdate(d *schema.ResourceData, m interface{}) error {
 		profile := populateUserProfile(d)
 		userBody := okta.User{Profile: profile}
 
-		_, _, err := client.User.UpdateUser(d.Id(), userBody)
+		_, _, err := client.User.UpdateUser(d.Id(), userBody, nil)
 		if err != nil {
 			return fmt.Errorf("[ERROR] Error Updating User in Okta: %v", err)
 		}
@@ -450,7 +450,7 @@ func ensureUserDelete(id, status string, client *okta.Client) error {
 	}
 
 	for i := 0; i < passes; i++ {
-		_, err := client.User.DeactivateOrDeleteUser(id)
+		_, err := client.User.DeactivateOrDeleteUser(id, nil)
 		if err != nil {
 			return fmt.Errorf("Failed to deprovision or delete user from Okta: %v", err)
 		}

--- a/okta/resource_user_schema.go
+++ b/okta/resource_user_schema.go
@@ -2,7 +2,6 @@ package okta
 
 import (
 	"fmt"
-	"time"
 
 	articulateOkta "github.com/articulate/oktasdk-go/okta"
 	"github.com/hashicorp/terraform/helper/schema"
@@ -133,18 +132,7 @@ func resourceUserSchemaRead(d *schema.ResourceData, m interface{}) error {
 	if err != nil {
 		return err
 	} else if subschema == nil {
-		// See https://github.com/articulate/terraform-provider-okta/issues/144
-		// Occassionally a schema prop would be created and the read would not find it.
-		// There appears to be a delay of availability on the Okta side, thus the backoff.
-		fmt.Println("Could not find an existence subschema property, backing off and retrying. Known timing issue")
-		time.Sleep(time.Second * 3)
-
-		subschema, err = getSubSchema(d, m)
-		if err != nil {
-			return err
-		} else if subschema == nil {
-			return fmt.Errorf("Okta did not return a subschema for \"%s\"", d.Id())
-		}
+		return fmt.Errorf("Okta did not return a subschema for \"%s\". This is a known bug caused by terraform running concurrent subschena resources. See https://github.com/articulate/terraform-provider-okta/issues/144. New resource is on its way to solve this limitation.", d.Id())
 	}
 
 	d.Set("array_type", subschema.Items.Type)

--- a/okta/user.go
+++ b/okta/user.go
@@ -351,14 +351,27 @@ func populateUserProfile(d *schema.ResourceData) *okta.UserProfile {
 	return &profile
 }
 
+func listUserOnlyRoles(c *okta.Client, userId string) (userOnlyRoles []*okta.Role, resp *okta.Response, err error) {
+	roles, resp, err := c.User.ListAssignedRoles(userId, nil)
+	if err != nil {
+		return
+	}
+
+	for _, role := range roles {
+		if role.AssignmentType == "USER" {
+			userOnlyRoles = append(userOnlyRoles, role)
+		}
+	}
+
+	return
+}
+
 func setAdminRoles(d *schema.ResourceData, c *okta.Client) error {
 	roleTypes := make([]interface{}, 0)
 
 	// set all roles currently attached to user in state
-	roles, resp, err := c.User.ListAssignedRoles(d.Id(), nil)
+	roles, resp, err := listUserOnlyRoles(c, d.Id())
 
-	// only super admin can list roles if this fails assume the user has no
-	// admin roles
 	if err != nil {
 		if resp.StatusCode == http.StatusForbidden {
 			log.Printf("[INFO] Insufficient permissions to get Admin Roles, skipping.")
@@ -447,7 +460,7 @@ func flattenUser(u *okta.User, d *schema.ResourceData) (map[string]interface{}, 
 
 // need to remove from all current admin roles and reassign based on terraform configs when a change is detected
 func updateAdminRolesOnUser(u string, r []string, c *okta.Client) error {
-	roles, _, err := c.User.ListAssignedRoles(u, nil)
+	roles, _, err := listUserOnlyRoles(c, u)
 
 	if err != nil {
 		return fmt.Errorf("[ERROR] Error Updating Admin Roles On User: %v", err)
@@ -510,7 +523,7 @@ func updateUserStatus(uid string, desiredStatus string, c *okta.Client) error {
 	case "SUSPENDED":
 		_, statusErr = c.User.SuspendUser(uid)
 	case "DEPROVISIONED":
-		_, statusErr = c.User.DeactivateUser(uid)
+		_, statusErr = c.User.DeactivateUser(uid, nil)
 	case "ACTIVE":
 		switch user.Status {
 		case "SUSPENDED":

--- a/okta/utils.go
+++ b/okta/utils.go
@@ -15,7 +15,6 @@ import (
 
 	articulateOkta "github.com/articulate/oktasdk-go/okta"
 	"github.com/hashicorp/terraform/helper/schema"
-	"github.com/okta/okta-sdk-golang/okta/cache"
 )
 
 func buildSchema(s, t map[string]*schema.Schema) map[string]*schema.Schema {
@@ -322,12 +321,7 @@ func getSupplementFromMetadata(meta interface{}) *ApiSupplement {
 }
 
 func getRequestExecutor(m interface{}) *okta.RequestExecutor {
-	c := m.(*Config)
-	config := okta.NewConfig().
-		WithOrgUrl(getBaseUrl(m)).
-		WithToken(c.apiToken).
-		WithCache(false)
-	return okta.NewRequestExecutor(nil, cache.NewNoOpCache(), config)
+	return getOktaClientFromMetadata(m).GetRequestExecutor()
 }
 
 func is404(status int) bool {

